### PR TITLE
Introduce CompositeProject

### PR DESCRIPTION
**Pull request by ghost**
fixes: https://github.com/sbt/sbt/issues/3042

/CC the sbt-crossproject "crew" - @sjrd @densh @MasseGuillaume

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @BennyHill,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by ghost** [comment]
Just did the cla....please shout if I found a way to mess it up

**Comment by ghost** [comment]
@sjrd Thanks for feedback...tbh, I PR'd early for such feedback.

I'll add you changes :+1:

**Comment by ghost** [comment]
Also, whilst I modified `BuildDef`, I'm not entirely certain that this is in fact (still) used. I'll continue to update both....

File: main/src/main/scala/sbt/Project.scala [hunk 1]
```diff
@@ -119,7 +119,13 @@ sealed trait ProjectDefinition[PR <: ProjectReference] {
     if (ts.isEmpty) Nil else s"$label: $ts" :: Nil
 }
 
-sealed trait Project extends ProjectDefinition[ProjectReference] {
+trait CompositeProject {
```
Comment (dwijnand) [inline main/src/main/scala/sbt/Project.scala:124]: could you seal this please?

**Comment by dwijnand** [comment]
I've retargeted this to be merged into 1.x because it's a bit more than just a straight bug fix.

however that's caused merge conflicts with unrelated changes already in 1.x - sorry!

File: main/src/main/scala/sbt/Project.scala (hunk 1 above)
Comment (ghost) [reply main/src/main/scala/sbt/Project.scala:124]: In the implementation of, say, `CrossProject`, we would want to extend this. In the sbt-test, I just create an instance, but in reality, I _think_ we would want to extend. 

Of course, I may have got the use wrong. But my impression was that by doing this, its then possible to have a `CrossProject` as a „real“ `Project`

Comment (ghost) [reply main/src/main/scala/sbt/Project.scala:124]: Hm... maybe I‘m wrong, sigh. OK 👍

**Comment by ghost** [comment]
Heh, I know - I went to change it before the PR....but saw the changes and as I was unsure of the right branch to use, just submitted. It was 50/50, never mind, really np 😄

File: main/src/main/scala/sbt/Project.scala (hunk 1 above)
Comment (sjrd) [reply main/src/main/scala/sbt/Project.scala:124]: Er ... @BennyHill is right, here. If it's sealed, this thing is completely useless ;) The whole point of this thing is that we can indeed have `CrossProject` implement that trait (which is a pure Java interface, btw). Otherwise, how do you propose we leverage this in sbt-crossproject?

Comment (ghost) [reply main/src/main/scala/sbt/Project.scala:124]: OK.... How about that when I do the other changes, I update the tests to mimick invisaged usage. That way, it^s clear-cut

Comment (dwijnand) [reply main/src/main/scala/sbt/Project.scala:124]: hmm I see your point.

what if we provide a `ProjectsList` subclass?

I'm not completely against the idea of an open interface, I just prefer to not have one.

Comment (sjrd) [reply main/src/main/scala/sbt/Project.scala:124]: > what if we provide a ProjectsList subclass?

What exactly does that buy you, if that one is open, compared to having the trait open? The only thing I see is that you would be able to add a concrete field in a binary compatible way, but ... that would be weird anyway.

For things that build on top of it, it's a limitation, as they cannot extend `ProjectsList` *and* another class, but they could extend `CompositeProject` and a class, because `CompositeProject` is an interface.

Comment (dwijnand) [reply main/src/main/scala/sbt/Project.scala:124]: > What exactly does that buy you, if that one is open

it wouldn't be open `final class ProjectsList` that extends `CompositeProject` I was thinking.

Comment (sjrd) [reply main/src/main/scala/sbt/Project.scala:124]: Then I don't see how I can use it in `CrossProject` either, since I need `crossProject` to return something that is both a `CompositeProject` (for sbt to pick it up as such) *and* have methods that are specific to `CrossProject`s (so that I can do `.dependsOn(foo.jvm)` in some other JVM project, at least. If my only options to have a `CompositeProject` are `Project` which is sealed and `ProjectsList` which is final, it's no good.

Basically, think about how I can declare `CrossProject` so that the following typechecks and is identified as a `CompositeProject`:
```scala
lazy val foo = crossProject
  .settings(...)
  .jvmSettings(...)
  .jsSettings(...)

lazy val bar = project
  .settings(...)
  .dependsOn(foo.jvm)
```

Comment (sjrd) [reply main/src/main/scala/sbt/Project.scala:124]: With an open `CompositeProject`, it's super easy. I just alter the [declaration of `CrossProject`](https://github.com/portable-scala/sbt-crossproject/blob/f1a6e1c7c18f109599174a4f8cfad60e1afb9a21/sbt-crossproject/src/main/scala/sbtcrossproject/CrossProject.scala#L10) with `extends CompositeProject` and
```scala
def componentProjects: Seq[Project] = projects.values.toSeq
```

Comment (dwijnand) [reply main/src/main/scala/sbt/Project.scala:124]: got it. thanks for detailing the use case.

**Comment by ghost** [comment]
just fyi, one more commit to come with a mock `CrossProject extends CompositeProject` ....

**Comment by ghost** [comment]
Actually, is this OK as is?

File: sbt/src/sbt-test/project/sbt-composite-projects/changes/shadow.sbt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+lazy val bar = project
+  .dependsOn(foo.jvm)
+
+val g = taskKey[Unit]("A task in the root project")
+g := println("Hello.")
+
+
+check := {
+  val verJvm = (version in foo.jvm).?.value
+  assert (verJvm == Some("0.2.0"))
```
Comment (eed3si9n) [inline sbt/src/sbt-test/project/sbt-composite-projects/changes/shadow.sbt:26]: Nice

**Review by eed3si9n — APPROVED** [review]
overall LGTM
could you add a release note in https://github.com/sbt/sbt/tree/1.x/notes/1.2.0 plz?

File: main/src/main/scala/sbt/Project.scala [hunk 3]
```diff
@@ -121,7 +121,34 @@ sealed trait ProjectDefinition[PR <: ProjectReference] {
...
+  def componentProjects: Seq[Project]
+}
+
+object CompositeProject {
```
Comment (sjrd) [inline main/src/main/scala/sbt/Project.scala]: This companion object can probably be `private[sbt]`. Users of sbt don't need to know about `uniqueId`.

File: main/src/main/scala/sbt/internal/EvaluateConfigurations.scala [hunk 4]
```diff
@@ -151,10 +151,13 @@ private[sbt] object EvaluateConfigurations {
...
       {
-        val projects =
-          definitions.values(loader).collect {
-            case p: Project => resolveBase(file.getParentFile, p)
+        val projects = {
+          val compositeProjects = definitions.values(loader).flatMap {
+            case p: CompositeProject => p.componentProjects.map(resolveBase(file.getParentFile, _))
+            case _                   => Nil
           }
+          CompositeProject.uniqueId(compositeProjects)
```
Comment (sjrd) [inline main/src/main/scala/sbt/internal/EvaluateConfigurations.scala]: If you first `flatMap`, *then* call `uniqueId`, you have already lost the distinction between things that were in the original list versus those added via the `componentProjects`. That can't properly work: they will *all* be kept.

File: main/src/main/scala/sbt/internal/BuildDef.scala [hunk 5]
```diff
@@ -17,7 +17,10 @@ import sbt.internal.inc.ReflectUtilities
 
 trait BuildDef {
   def projectDefinitions(@deprecated("unused", "") baseDirectory: File): Seq[Project] = projects
-  def projects: Seq[Project] = ReflectUtilities.allVals[Project](this).values.toSeq
+  def projects: Seq[Project] =
+    CompositeProject.uniqueId(
+      ReflectUtilities.allVals[CompositeProject](this).values.toSeq.flatMap(_.componentProjects))
```
Comment (sjrd) [inline main/src/main/scala/sbt/internal/BuildDef.scala]: Same problem that I mentioned in the other comment, obviously.

File: main/src/main/scala/sbt/Project.scala [hunk 6]
```diff
@@ -121,7 +121,34 @@ sealed trait ProjectDefinition[PR <: ProjectReference] {
...
+
+  /**
+   *  If two projects with the same id appear in `compositeProjects` and
+   *  in `compositeProjects.map(_.componentProjects)`, the one in
+   *  `compositeProjects` wins.
+   *  This is necessary for backward compatibility with the idiom:
+   *    lazy val foo = crossProject
+   *    lazy val fooJVM = foo.jvm.setting
+   */
+  def uniqueId(compositeProjects: Seq[Project]): Seq[Project] = {
```
Comment (sjrd) [inline main/src/main/scala/sbt/Project.scala]: Basically, this method should take the original `Seq[CompositeProject]`, not a `Seq[Project]`. If it takes the `Seq[Project]`, it's too late: you don't know what were the projects that were in the initial `compositeProjects` versus those that were added through `componentProjects`.

File: main/src/main/scala/sbt/internal/EvaluateConfigurations.scala (hunk 4 above)
Comment (sjrd) [reply main/src/main/scala/sbt/internal/EvaluateConfigurations.scala]: It should be possible to *first* call `CompositeProject.expand(definitions.values(loader))`, then `map` the result with `.map(resolveBase(file.getParentFile, _))`.

File: main/src/main/scala/sbt/Project.scala (hunk 6 above)
Comment (sjrd) [reply main/src/main/scala/sbt/Project.scala]: Hence, I would recommend calling this method `expand`, not `uniqueId`, since its job would be to *expand* the list of `CompositeProject` into the appropriate list of `Project`s.

File: sbt/src/sbt-test/project/sbt-composite-projects/changes/shadow.sbt [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+// Based on sbt-file-projects test
+lazy val foo = new CompositeProject
+{
+  val jvm = Project.apply("jvm", new File("jvm"))
+  val js = Project.apply("js", new File("js"))
+  def componentProjects: Seq[Project] = Seq(jvm, js)
+}
+
+lazy val fooJVM = foo.jvm.settings(version := "0.2.0") // this one needs to win
```
Comment (sjrd) [inline sbt/src/sbt-test/project/sbt-composite-projects/changes/shadow.sbt]: This is very good.

To make it even better, you can also declare the `lazy val fooJS`, but *above* the `lazy val foo` (they're lazy vals, so it will work). But that should ensure that the shadowing treatment does not accidentally work *via* the ordering (which I suspect is currently the case).

**Review by sjrd — CHANGES_REQUESTED** [review]
I have a series of comments, but basically all related to the same root cause: I think the refactoring you applied with `uniqueId` gives the wrong answer.

**Review by eed3si9n — DISMISSED** [review]
Yea. @sjrd is right about the `uniqueId`.

**Comment by ghost** [comment]
Thanks @sjrd for the thorough review, feedback and suggestions. I will do these tomorrow with fresh eyes, along with the release notes

File: sbt/src/sbt-test/project/sbt-composite-projects/changes/shadow.sbt (hunk 7 above)
Comment (ghost) [reply sbt/src/sbt-test/project/sbt-composite-projects/changes/shadow.sbt]: >  (which I suspect is currently the case).

Indeed, adding a `lazy val fooJS` first does fail...

File: main/src/main/scala/sbt/Project.scala [hunk 8]
```diff
@@ -121,7 +121,44 @@ sealed trait ProjectDefinition[PR <: ProjectReference] {
...
+   *    lazy val fooJVM = foo.jvm.settings(...)
+   *  }}}
+   *  and the rarer:
+   *  {{{
+   *    lazy val fooJS = foo.js.settings(...)
+   *    lazy val foo = crossProject
+   *    lazy val fooJVM = foo.jvm.settings(...)
+   *  }}}
+   */
+  def expand(projects: Seq[Project], compositeProjects: Seq[CompositeProject]): Seq[Project] = {
```
Comment (sjrd) [inline main/src/main/scala/sbt/Project.scala]: You can compute `projects` inside `expand` from `compositeProjects` once and for all, since all projects are also composite projects, so that call sites don't need to know about this:
```scala
val projects = compositeProjects.collect {
  case project: Project => project
}
```
No need to pass it as an additional parameter when `expand` can compute it itself, especially since the call sites are basically `collect` ing as well to give it to `expand` anyway.

**Review by sjrd — APPROVED** [review]
:)

File: main/src/main/scala/sbt/Project.scala (hunk 8 above)
Comment (ghost) [reply main/src/main/scala/sbt/Project.scala]: Yep, so I've removed it - thx :+1:

**Review by dwijnand — APPROVED** [review]
